hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/S3GuardFsck.java

```diff
@@ -0,0 +1,290 @@
+    queue.add(root);
+
+    while (!queue.isEmpty()) {
+      // pop front node from the queue
+      final S3AFileStatus currentDir = queue.poll();
+
+      // Get a listing of that dir from s3 and add just the files.
+      // (Each directory will be added as a root.)
+      // Files should be casted to S3AFileStatus instead of plain FileStatus
+      // to get the VersionID and Etag.
+      final Path currentDirPath = currentDir.getPath();
+      // TODO Do we need to do a HEAD for each children in the path if we
+      //  want the versionID? In the listing it is empty.
+      final List<S3AFileStatus> children =
+          Arrays.asList(rawFS.listStatus(currentDirPath)).stream()
+              .filter(status -> !status.isDirectory())
+              .map(S3AFileStatus.class::cast).collect(toList());
+
+      comparePairs.addAll(
+          compareS3DirToMs(currentDir, children).stream()
+              .filter(comparePair -> comparePair.containsViolation())
+              .collect(Collectors.toList())
+      );
+
+      // Add each dir to queue
+      children.stream().filter(pm -> pm.isDirectory())
+          .forEach(pm -> queue.add(pm));
+    }
+
+    // Create a handler and handle each violated pairs
+    S3GuardFsckViolationHandler handler =
+        new S3GuardFsckViolationHandler(rawFS, metadataStore);
+    comparePairs.forEach(handler::handle);
+  }
+
+  private List<ComparePair> compareS3DirToMs(S3AFileStatus s3CurrentDir,
+      List<S3AFileStatus> children) throws IOException {
+    final Path path = s3CurrentDir.getPath();
+    final PathMetadata pathMetadata = metadataStore.get(path);
+    List<ComparePair> violationComparePairs = new ArrayList<>();
+
+    final ComparePair rootComparePair =
+        compareFileStatusToPathMetadata(s3CurrentDir, pathMetadata);
+    if (rootComparePair.containsViolation()) {
+      violationComparePairs.add(rootComparePair);
+    }
+
+    children.forEach(s3ChildMeta -> {
+      try {
+        final PathMetadata msChildMeta =
+            metadataStore.get(s3ChildMeta.getPath());
+        final ComparePair comparePair =
+            compareFileStatusToPathMetadata(s3ChildMeta, msChildMeta);
+        if (comparePair.containsViolation()) {
+          violationComparePairs.add(comparePair);
+        }
+      } catch (Exception e) {
+        e.printStackTrace();
+      }
+    });
+
+    return violationComparePairs;
+  }
+
+  private ComparePair compareFileStatusToPathMetadata(S3AFileStatus s3FileStatus,
+      PathMetadata msPathMetadata) throws IOException {
+    final Path path = s3FileStatus.getPath();
+    System.out.println("== Path: " + path);
+    ComparePair comparePair = new ComparePair(s3FileStatus, msPathMetadata);
+
+    if (!path.equals(path(ROOT_PATH_STRING))) {
+      final Path parentPath = path.getParent();
+      final PathMetadata parentPm = metadataStore.get(parentPath);
+
+      if (parentPm == null) {
+        comparePair.violations.add(Violation.NO_PARENT_ENTRY);
```

Comment:
I think this approach is okay. I had pictured it being more pluggable and modular, but I would imagine that would be woefully inefficient. I think we just need to make sure this code stays as clean & readable as it is, and that will get harder over time as we add new problems we come across.